### PR TITLE
[IMP]tests: autoretry mechanism for stagings

### PR DESCRIPTION
**Pull request by Xavier-Do**
This pr proposes an auto retry mechanism for tests. This shouldn't impact normal testing: tests are not supposed to fail, but the growing number of tests and pull requests can lead to some bottleneck when a staging fails because of a random error. This mechanism should help to reduce splits/the need to retry a failed pr. 

This branch have been tested with the nightly multi build, creating 40 identical build without test-tags to disable know random errors. This multi build is used to detect test failing randomly, this is an excellent candidate to detect the effect of the retry. 
On average, with the current base of this pull request, there is between 10 en 15 failures over 40 build. 
With the auto retry mechanism, only 1 build failed over 40 builds since the same error was triggered twice: https://runbot.odoo.com/runbot/build/9972771. This is simply because with the retry mechanism, an error that has a probability of p to fail randomly will still have a probability of p² to fail with the retry mechanism. A error that occurs 10% of the time should only appear 1% of the time with one retry.  In most of the case, the retry is sucessfull: https://runbot.odoo.com/runbot/build/10053257 

The current solution to allow to enable this mechanism only in some cases (staging) is to check an environment variable "ODOO_TEST_FAILURE_RETRIES" that defines a number of retry. This will allow to retry more than once if an error still occurs to ofen with the autoretry. 

**Random error:** 
image
**Real error:** 
image

The mechanism will run multiple time the same test on the same test_case, meaning that some modification on self may impact the second execution. The following code is an example of how this could be problematic, but also a good example to test the auto-retry mechanism.

```python

class TestRetry(HttpCase):
    def test_fail(self):
        self.t = getattr(self, 't', 0) + 1
        if True or self.t == 1:
            import logging
            _logger = logging.getLogger('test_a')

            with self.assertLogs(level="ERROR"):
                _logger.error("This shouldn't be log at all")

            with mute_logger('test_a'):
                _logger.error("This shouldn't be logged (mute)")
    
            _logger.error("This should be log")
```

image

As we can see here the error logs are also managed, and emit at a lower level the first time, butany log higher than 25 will make the test "failed" and the autoretry mechanism will be triggered. The second time, everything is logged normally. We also need to replace Traceback by _Traceback to avoid being catched by runbot Traceback detection regexes. 

The inspiration here commes from the assertLogs, that replace all handlers. The mute_logger had to be adapted to use the same strategy, so that quite_logger won't detect logs catched by mute_logger or assertLogs.

### CODE REVIEW

File: odoo/tests/common.py [hunk 1]
```diff
@@ -348,6 +348,22 @@ def __init__(self, methodName='runTest'):
         self.addTypeEqualityFunc(etree._Element, self.assertTreesEqual)
         self.addTypeEqualityFunc(html.HtmlElement, self.assertTreesEqual)
 
+    def run(self, result):
+        ODOO_TEST_FAILURE_RETRIES = int(os.environ.get('ODOO_TEST_FAILURE_RETRIES', 1))
+        failure = False
+        for retry in range(ODOO_TEST_FAILURE_RETRIES):
+            if retry == 0 or failure:
```
Comment (xmo-odoo) [inline odoo/tests/common.py:356]: This seems not useful: if `not failure` then the loop is broken out of, so we should only be able to reach `retry != 0` if `failure`.

File: odoo/tests/common.py [hunk 2]
```diff
@@ -348,6 +348,22 @@ def __init__(self, methodName='runTest'):
...
+                if retry:
+                    _logger.runbot(f'Retrying a failed test: {self}')
+                if retry != ODOO_TEST_FAILURE_RETRIES-1:
```
Comment (xmo-odoo) [inline odoo/tests/common.py:359]: Would be clearer with the test reversed.

File: odoo/tests/common.py [hunk 3]
```diff
@@ -348,6 +348,22 @@ def __init__(self, methodName='runTest'):
         self.addTypeEqualityFunc(etree._Element, self.assertTreesEqual)
         self.addTypeEqualityFunc(html.HtmlElement, self.assertTreesEqual)
 
+    def run(self, result):
+        ODOO_TEST_FAILURE_RETRIES = int(os.environ.get('ODOO_TEST_FAILURE_RETRIES', 1))
```
Comment (xmo-odoo) [inline odoo/tests/common.py:352]:
```suggestion
        ODOO_TEST_FAILURE_RETRIES = int(os.environ.get('ODOO_TEST_FAILURE_RETRIES', 0)) + 1
```
seems more logical, if somebody sets `ODOO_TEST_FAILURE_RETRIES=0` they want the tests to not be retried (run up to once) and `ODOO_TEST_FAILURE_RETRIES=1` they want the tests to be retried once (run up to twice).

And the local should then be renamed to something more like `tests_run_count`.

File: odoo/tests/common.py [hunk 4]
```diff
@@ -348,6 +348,23 @@ def __init__(self, methodName='runTest'):
...
+        failure = False
+        ODOO_TEST_FAILURE_RETRIES = 2 # TODO remove me
+        for retry in range(ODOO_TEST_FAILURE_RETRIES):
+            if retry == 0 or failure:
+                if retry:
+                    _logger.runbot(f'Retrying a failed test: {self}')
+                if retry != ODOO_TEST_FAILURE_RETRIES-1:
+                    with result.soft_fail(), quiet_logger() as quiet_log:
+                        super().run(result)
+                        failure = result.had_failure or quiet_log.had_error_log
```
Comment (xmo-odoo) [inline odoo/tests/common.py:362]: > `or quiet_log.had_error_log`

I still don't think I agree with that: though it may (will?) fail the build on runbot, `_logger.error` does not fail the tests normally.

File: odoo/tests/runner.py [hunk 5]
```diff
@@ -18,10 +19,22 @@ def __init__(self):
...
         self.queries_start = None
+        self._soft_fail = False
+        self.had_failure = False
 
     def __str__(self):
         return f'{len(self.failures)} failed, {len(self.errors)} error(s) of {self.testsRun} tests'
 
+    @contextlib.contextmanager
+    def soft_fail(self):
+        self.had_failure = False
```
Comment (xmo-odoo) [inline odoo/tests/runner.py:30]: Is touching this really necessary given `startTest` will automatically update it?

File: odoo/tools/misc.py [hunk 6]
```diff
@@ -781,19 +781,26 @@ def do_stuff():
...
 
     def __enter__(self):
-        for logger in self.loggers:
-            assert isinstance(logger, str),\
-                "A logger name must be a string, got %s" % type(logger)
-            logging.getLogger(logger).addFilter(self)
+        for logger_name in self.loggers:
+            logger = logging.getLogger(logger_name)
+            self.old_handlers[logger_name] = logger.handlers[:]
+            self.old_propagate[logger_name] = logger.propagate
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:795]: Seems unnecessary to have two different maps for what's in essence a logger's configuration. This could be a single entry in a single map. *Copying* the list also seems unnecessary since we're replacing it afterwards (not clearing and updating it).

And since we're messing with the low-level logging information, this should probably be done while holding the logging lock.

File: odoo/tools/misc.py [hunk 7]
```diff
@@ -781,19 +781,26 @@ def do_stuff():
     """
     def __init__(self, *loggers):
         self.loggers = loggers
+        self.old_handlers = {}
+        self.old_propagate = {}
+        super().__init__()
 
     def filter(self, record):
         return 0
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:789]: Seems unnecessary if this is now a handler.

File: odoo/tools/misc.py [hunk 8]
```diff
@@ -802,6 +809,44 @@ def deco(*args, **kwargs):
                 return func(*args, **kwargs)
         return deco
 
+    def emit(self, record):
+        pass
+
+
+class quiet_logger(logging.Handler):
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:816]: The naming is confusing, this is neither a "quiet logger" nor quite something which quiets a specific (named) logger (like `mute_logger`).

File: odoo/tools/misc.py [hunk 9]
```diff
@@ -802,6 +809,44 @@ def deco(*args, **kwargs):
...
+        if record.levelno > 25:
+            record.levelname = f'_{record.levelname}'
+            record.levelno = 25
+            self.had_error_log = True
+
+        for handler in self.old_handlers:
+            handler.emit(record)
+
+
+
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:849]: You're also getting into negative space?

File: odoo/tools/misc.py [hunk 10]
```diff
@@ -802,6 +809,44 @@ def deco(*args, **kwargs):
                 return func(*args, **kwargs)
         return deco
 
+    def emit(self, record):
+        pass
+
+
+class quiet_logger(logging.Handler):
+    """Temporary lower the max logging level.
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:817]: Only really works if there are no handlers outside the root loggers though.

File: odoo/tools/misc.py [hunk 11]
```diff
@@ -802,6 +809,44 @@ def deco(*args, **kwargs):
...
+    """
+    def __init__(self):
+        self.old_handlers = None
+        self.old_propagate = None
+        self.had_error_log = False
+        super().__init__()
+
+    def __enter__(self):
+        logger = logging.getLogger()
+        self.old_handlers = logger.handlers[:]
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:827]: Since this is manipulating the handlers directly, it could just `logger.handlers.insert(0, self)` no?

File: odoo/tools/misc.py [hunk 12]
```diff
@@ -802,6 +809,44 @@ def deco(*args, **kwargs):
...
+        self.old_propagate = logger.propagate
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:828]: Don't think this is necessary or useful, the root logger has no parent so there can't be any propagation from it.

File: odoo/tools/misc.py [hunk 13]
```diff
@@ -802,6 +809,44 @@ def deco(*args, **kwargs):
...
+
+
+class quiet_logger(logging.Handler):
+    """Temporary lower the max logging level.
+    """
+    def __init__(self):
+        self.old_handlers = None
+        self.old_propagate = None
+        self.had_error_log = False
+        super().__init__()
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:823]: Should call `super()` first unless there are very good reasons not to.

**Comment by Xavier-Do** [comment]
Looks like a lot of log 25 are produced, need to fix this.

File: odoo/tools/misc.py (hunk 10 above)
Comment (Xavier-Do) [reply odoo/tools/misc.py:811]: Good enough for us right now.

**Review by d-fence — APPROVED** [review]
Good job. Looks good to me.
That's a great improvement for the testing tools.

File: odoo/tools/misc.py [hunk 14]
```diff
@@ -802,6 +803,45 @@ def deco(*args, **kwargs):
                 return func(*args, **kwargs)
         return deco
 
+    def emit(self, record):
+        pass
+
+
+class quiet_logging(logging.Handler):
```
Comment (rco-odoo) [inline odoo/tools/misc.py:810]: What about `lower_logging` instead, since it is what it does?
Also, I would pass the maximal logging level as a parameter, instead of hardcoding `25`.

File: odoo/tests/common.py [hunk 15]
```diff
@@ -348,6 +348,22 @@ def __init__(self, methodName='runTest'):
...
+        tests_run_count = int(os.environ.get('ODOO_TEST_FAILURE_RETRIES', 0)) + 1
+        failure = False
+        tests_run_count = 2 # TODO remove me
+        for retry in range(tests_run_count):
+            if retry:
+                _logger.runbot(f'Retrying a failed test: {self}')
+            if retry < tests_run_count-1:
+                with result.soft_fail(), quiet_logging() as quiet_log:
+                    super().run(result)
+                    failure = result.had_failure or quiet_log.had_error_log
```
Comment (rco-odoo) [inline odoo/tests/common.py:361]: What about using an alternative object for `result` instead?  Something like
```python
local_result = SoftResult(result)       # behaves as a proxy to result
with lower_logging(25) as log:
    super().run(local_result)
    failure = local_result.has_failed or log.has_error
```
This would avoid plumbing `OdooTestResult` with a feature that you generally don't use.

File: odoo/tools/misc.py [hunk 16]
```diff
@@ -767,7 +767,7 @@ def __missing__(self, key):
         return unquote(key)
 
 
-class mute_logger(object):
+class mute_logger(logging.Handler):
```
Comment (rco-odoo) [inline odoo/tools/misc.py:770]: Why did you change this?
I don't understand.

File: odoo/tests/common.py [hunk 17]
```diff
@@ -348,6 +348,22 @@ def __init__(self, methodName='runTest'):
         self.addTypeEqualityFunc(etree._Element, self.assertTreesEqual)
         self.addTypeEqualityFunc(html.HtmlElement, self.assertTreesEqual)
 
+    def run(self, result):
+        tests_run_count = int(os.environ.get('ODOO_TEST_FAILURE_RETRIES', 0)) + 1
+        failure = False
+        tests_run_count = 2 # TODO remove me
```
Comment (rco-odoo) [inline odoo/tests/common.py:354]: Don't forget to remove this one! :wink:

Comment (Xavier-Do) [reply odoo/tests/common.py:354]: Yes sure ;)

File: odoo/tools/misc.py (hunk 16 above)
Comment (Xavier-Do) [reply odoo/tools/misc.py:770]: I gave a quick explanation about that at the end of the in the pr message but I either don't remember the precise corner case or it's not usefull anymore...

```python
    with lower_logging(25):
        with assertLogs(level='ERROR'):
            logger.error()  
```
In short, to work with assertLogs, lower_logging must be an handler. And during my testing I needed to change mute_logger to and handler too for some reason, but I'll try to remember why exactly or revert this change.

**Comment by Xavier-Do** [comment]
@robodoo override=ci/security

Using os to access env variables.

**Comment by Xavier-Do** [comment]
@robodoo r+
